File: 024/swap-nodes-in-pairs.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	dummyHead := &ListNode{-1, head}

	var loop func(*ListNode)
	loop = func(node *ListNode) {
		if node != nil && node.Next != nil && node.Next.Next != nil {
			nextNode := node.Next
			node.Next = nextNode.Next
			nextNode.Next = nextNode.Next.Next
			node.Next.Next = nextNode
			loop(nextNode)
		} else {
			return
		}
	}

	loop(dummyHead)
	return dummyHead.Next
}

File: 024/SwapNodesInPairs.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int = 0) {
 *     var next: ListNode? = null
 * }
 */
class SwapNodesInPairs {
    fun swapPairs(head: ListNode?): ListNode? {
        var dummyHead = ListNode()
        dummyHead.next = head
        
        fun loop(node: ListNode?) {
            if (node != null && node.next != null && node.next.next != null) {
                var nextNode = node.next
                node.next = nextNode.next
                nextNode.next = nextNode.next.next
                node.next.next = nextNode
                loop(nextNode)
            } else return
        }
        
        loop(dummyHead)
        return dummyHead.next
    }
}

class SwapNodesInPairs {
    fun swapPairs(head: ListNode?): ListNode? {
        if (head == null) return null
        val second = head.next ?: return head
        head.next = swapPairs(second.next)
        second.next = head
        return second
    }
}